src/libled.c

```diff
@@ -0,0 +1,456 @@
+		return LED_STATUS_OUT_OF_MEMORY;
+	}
+	if (!list_append(&result->slot_list, s)) {
+		free(s);
+		list_erase(&result->slot_list);
+		return LED_STATUS_OUT_OF_MEMORY;
+	}
+	return LED_STATUS_SUCCESS;
+}
+
+led_status_t led_slots_get(struct led_ctx *ctx)
+{
+	led_status_t status;
+
+	struct pci_slot *slot;
+
+	struct cntrl_device *ctrl_dev;
+	struct enclosure_device *encl = NULL;
+	char slot_id[PATH_MAX];
+
+	list_for_each(sysfs_get_pci_slots(ctx), slot) {
+		status = get_slot(ctx, pci_get_slot, LED_CNTRL_TYPE_VMD,
+					slot->sysfs_path, &ctx->sl);
+		if (LED_STATUS_SUCCESS != status)
+			return status;
+	}
+
+	list_for_each(sysfs_get_cntrl_devices(ctx), ctrl_dev) {
+		if (ctrl_dev->cntrl_type != LED_CNTRL_TYPE_NPEM)
+			continue;
+
+		status = get_slot(ctx, npem_get_slot, LED_CNTRL_TYPE_NPEM,
+					ctrl_dev->sysfs_path, &ctx->sl);
+		if (LED_STATUS_SUCCESS != status)
+			return status;
+	}
+
+	list_for_each(sysfs_get_enclosure_devices(ctx), encl) {
+		for (int i = 0; i < encl->slots_count; i++) {
+			snprintf(slot_id, PATH_MAX, "%s/%d", encl->dev_path, encl->slots[i].index);
+			status = get_slot(ctx, enclosure_get_slot, LED_CNTRL_TYPE_SCSI, slot_id,
+						&ctx->sl);
+			if (LED_STATUS_SUCCESS != status)
+				return status;
+		}
+	}
+
+	return LED_STATUS_SUCCESS;
+}
+
+led_status_t led_slot_get(struct led_ctx *ctx, enum led_cntrl_type cntrl, char *device, char *slot,
+				enum led_ibpi_pattern *state)
+{
+	struct slot_response slot_resp;
+	led_status_t status;
+
+	slot_response_init(&slot_resp);
+	switch (cntrl) {
+	case LED_CNTRL_TYPE_VMD:
+	{
+		status = pci_get_slot(ctx, device, slot, &slot_resp);
+		if (LED_STATUS_SUCCESS != status)
+			return status;
+		*state = slot_resp.state;
+		break;
+	}
+	case LED_CNTRL_TYPE_NPEM:
+	{
+		status = npem_get_slot(ctx, device, slot, &slot_resp);
+		if (LED_STATUS_SUCCESS != status)
+			return status;
+		*state = slot_resp.state;
+		break;
+	}
+	case LED_CNTRL_TYPE_SCSI:
+	{
+		status = enclosure_get_slot(ctx, device, slot, &slot_resp);
+		if (LED_STATUS_SUCCESS != status)
+			return status;
+		*state = slot_resp.state;
+		break;
+	}
+	default:
+		return LED_STATUS_NOT_SUPPORTED;
+	}
+
+	return LED_STATUS_SUCCESS;
+}
+
+led_status_t led_slot_set(struct led_ctx *ctx, enum led_cntrl_type cntrl, char *slot,
+				enum led_ibpi_pattern state)
+{
+	switch (cntrl) {
+	case LED_CNTRL_TYPE_VMD:
+	{
+		return pci_set_slot(ctx, slot, state);
+	}
+	case LED_CNTRL_TYPE_NPEM:
+	{
+		return npem_set_slot(ctx, slot, state);
+	}
+	case LED_CNTRL_TYPE_SCSI:
+	{
+		return enclosure_set_slot(ctx, slot, state);
+	}
+	default:
+		return LED_STATUS_NOT_SUPPORTED;
+	}
+}
+
+bool led_controller_slot_support(enum led_cntrl_type cntrl)
+{
+	return (cntrl == LED_CNTRL_TYPE_NPEM || cntrl == LED_CNTRL_TYPE_SCSI || cntrl == LED_CNTRL_TYPE_VMD);
+}
+
+led_status_t led_slot_list_next(struct led_ctx *ctx, struct led_slot_list_entry **slot_entry)
+{
+	if (!ctx->sl.iter) {
+		led_status_t rc = led_slots_get(ctx);
```

Comment:
> I think that we should create this list when we are creating led context, not when first request comes. We can make it configurable because I expect that slots are more advanced usecase not needed by default. It will safe some memory. What do you think?

I implemented it this way for the very reason that if you don't use it, you don't pay a memory penalty.  If we do it when we create the context, the memory is consumed regardless if an API user is intending on iterating on the list of slots or controllers.